whd/dispatcher.py

```diff
@@ -213,6 +219,130 @@ def _set_pr_labels(self, pr_event, resources):
+            logger.warning(
+                f"no resource versions found for pipeline {resource.pipeline.name} "
+                f"resource {resource.name} and PR {pr_event.number()}"
+            )
+            return
+
+        jobs_with_triggering_resource = [
+            job for job in resource.pipeline.jobs if job.is_triggered_by_resource(resource.name)
+        ]
+
+        # check if there is a pr resource version which was not build, if so return
+        # it with the corresponding job
+        for job in jobs_with_triggering_resource:
+            builds = concourse_api.job_builds(job.pipeline.name, job.name)
+            for resource_version in self._resource_version_ids_not_built(
+                resource_versions=pr_resource_versions,
+                builds=builds
+            ):
+                yield (job, resource, resource_version)
+
+    def _ensure_pull_requests_jobs_are_run(self, concourse_api, matching_resources, pr_event):
+        '''
+        check all PR resource versions for which no build exists. If a resource version has
+        no build, pin this resource version, trigger the corresponding job and unpin the resource
+        '''
+        for job, resource, resource_version in self._resource_versions_and_jobs(
+            concourse_api, matching_resources, pr_event
+        ):
+            logger.info(
+                f"Pin resource version {resource_version.version()} "
+                f"and trigger build for job {job.name}"
+            )
+            self._trigger_build_with_pinned_resource_version(
+                job=job,
+                resource_name=resource.name,
+                resource_version_id=resource_version.id()
+            )
+
+    def _resource_version_ids_not_built(self, resource_versions, builds):
+        '''
+        yield resource versions which have not been built by checking if the resource version
+        reference is found in an existing build
+        '''
+        for resource_version in resource_versions:
+            resource_version_in_build = False
+            for build in reversed(builds[-30:]):
+                build_plan = build.plan()
+                if build_plan.contains_version_ref(resource_version.version()['ref']):
+                    resource_version_in_build = True
+                    break
+            if not resource_version_in_build:
+                yield resource_version
+
+    def _trigger_build_with_pinned_resource_version(self, job, resource_name, resource_version_id):
+        '''
+        pin the passed resource version, trigger the corresponding job and unpin resourcce
+        version again
+        '''
+        def _job_running(
+            pipeline_name, job_name, build_number, concourse_api, retries=10, sleep_sec=3
+        ):
+            time.sleep(sleep_sec)
+            retries -= 1
+            if retries < 0:
+                logger.warning(
+                    f"Job {job_name} with build number {build_number} cannot be started"
```

Comment:
nit: _could_ not be started
using present tense indicates to me that it was apparent from the start that triggering is not possible (for whatever reason). what actually happened is that we failed to trigger the job after several attempts (which might be worth mentioning)